processor/src/host/debug.rs

```diff
@@ -1,52 +1,234 @@
+        // un-printed items
+        let i = num_items - 1;
+        if num_items == stack.len() {
+            println!("└── {i:>2}: {}\n", stack[i]);
+        } else {
+            println!("├── {i:>2}: {}", stack[i]);
+            println!("└── ({} more items)\n", stack.len() - num_items);
+        }
+    }
+
+    /// Prints the whole memory state at the cycle `clk` in context `ctx`.
+    fn print_mem_all<S: ProcessState>(&self, process: &S) {
+        let mem = process.get_mem_state(self.ctx);
+        let padding =
+            mem.iter().fold(0, |max, value| word_elem_max_len(Some(value.1)).max(max)) as usize;
+
+        println!("Memory state before step {} for the context {}:", self.clk, self.ctx);
+
+        // print the main part of the memory (wihtout the last value)
+        for (addr, value) in mem.iter().take(mem.len() - 1) {
+            print_mem_address(*addr as u32, Some(*value), false, false, padding);
+        }
+
+        // print the last memory value
+        if let Some((addr, value)) = mem.last() {
+            print_mem_address(*addr as u32, Some(*value), true, false, padding);
+        }
+    }
+
+    /// Prints memory values in the provided addresses interval.
+    fn print_mem_interval<S: ProcessState>(&self, process: &S, n: u32, m: u32) {
+        let mut mem_interval = Vec::new();
+        for addr in n..m + 1 {
+            mem_interval.push((addr, process.get_mem_value(self.ctx, addr)));
+        }
+
+        if n == m {
+            println!(
+                "Memory state before step {} for the context {} at address {}:",
+                self.clk, self.ctx, n
+            )
+        } else {
+            println!(
+                "Memory state before step {} for the context {} in the interval [{}, {}]:",
+                self.clk, self.ctx, n, m
+            )
+        };
+
+        print_interval(mem_interval, false);
+    }
+
+    /// Prints locals in provided indexes interval.
+    fn print_local_interval<S: ProcessState>(
+        &self,
+        process: &S,
+        interval: (u32, u32),
+        num_locals: u32,
+    ) {
+        let mut local_mem_interval = Vec::new();
+        let local_memory_offset = self.fmp - num_locals + 1;
+
+        // in case start index is 0 and end index is 2^16, we should print all available locals.
+        let (start, end) = if interval.0 == 0 && interval.1 == u16::MAX as u32 {
+            (0, num_locals - 1)
+        } else {
+            interval
+        };
+        for index in start..end + 1 {
+            local_mem_interval
+                .push((index, process.get_mem_value(self.ctx, index + local_memory_offset)))
+        }
+
+        if interval.0 == 0 && interval.1 == u16::MAX as u32 {
+            println!("Local state before step {} for the context {}:", self.clk, self.ctx)
+        } else if interval.0 == interval.1 {
+            println!(
+                "Local state before step {} for the context {} at index {}:",
```

Comment:
I would change the printed out text to: "State of procedure local at index {} before step {}:".